compare-bug-version-and-git-branch.sh

```diff
@@ -53,29 +56,31 @@ while [ -z "${BUG_PRODUCT}" -a ${BZQTRY} -le 3 ]; do
     BUG_STATUS=$(cut -d: -f4 <<< "${BZQOUT}")
 done
 
-if [ "${BUG_STATUS}" != "NEW" -a "${BUG_STATUS}" != "POST" -a "${BUG_STATUS}" != "ASSIGNED" ]; then
-    echo "BUG id ${BUG} has an invalid status as ${BUG_STATUS}. Acceptable status values are NEW, ASSIGNED or POST."
+if [ "${BUG_PRODUCT}" != "GlusterFS" ]; then
+    echo "BUG id ${BUG} belongs to '${BUG_PRODUCT}' and not 'GlusterFS'."
     exit 1
 fi
 
-if [ "${BUG_GROUPS}" != '[]' ]; then
-    echo "BUG id ${BUG} is marked private, please remove the groups."
+if [ "${BUG_STATUS}" != "NEW" -a "${BUG_STATUS}" != "POST" -a "${BUG_STATUS}" != "ASSIGNED" -a "${BUG_STATUS}" != "MODIFIED" ]; then
+    echo "BUG id ${BUG} has an invalid status as ${BUG_STATUS}. Acceptable status values are NEW, ASSIGNED, POST or MODIFIED"
     exit 1
 fi
 
-if [ "${BUG_PRODUCT}" != "GlusterFS" ]; then
-    echo "BUG id ${BUG} belongs to '${BUG_PRODUCT}' and not 'GlusterFS'."
+if [ "${BUG_GROUPS}" != '[]' ]; then
+    echo "BUG id ${BUG} is marked private, please remove the groups."
     exit 1
 fi
 
 if [ "${GERRIT_BRANCH}" = "master" ]; then
     if [ "${BUG_VERSION}" != 'mainline' -a \
-         "${BUG_VERSION}" != 'pre-release' ]; then
-            echo "Change filed against the '${GERRIT_BRANCH} branch, but the BUG id is for '${BUG_VERSION}'"
-            exit 1
+                          "${BUG_VERSION}" != 'pre-release' ]; then
+        # This is Treated fine, because any user files a bug on release branch,
+        # but developers fix the issue on master first. Why bring one more step.
+        # Handle it properly in scripts.
```

Comment:
So does this mean we do not clone bugs into master? Meaning, if branch is 'master' and bug version is not against 'master' then if we allow the commit to go through, then closing bugs on a release basis will denote bad data.

So same Bug# will appear in 2 commits to different branches, once in master and once in the commit backport to the release branch.

When using Bugzilla, I think we should enforce the clone to master, even when users file bugs against releases (which I agree as per the comment would be the norm).

Reply:
I personally believe for one bug filed by user, if there is a fix, it would land in master first, so, it shouldn't be cloned. Only when there is a need for backport, we can clone it. What it means is, when a user files a bug, it will always gets fixed in next possible release. Let me know if that is fine.

Reply:
Here is a situation that I envisage with this change:
- Assume a bug was filed against 3.12
- Fixed using that bug ID in mainline
- Release 4.1 happens before the bug is backported to 3.12 and contains the fix due to it being branches from mainline

At this point the bug will get an update that it is closed and is available with 4.1 release. The bug was filed against 3.12 and when the backport and release for that happens it will get another update that it is fixed for 4.1.

Now, we call out version in which the bug is fixed in our BZ update text, so confusion is lesser, but the bug getting closed because the fix is available in a later release can be confusing.

It can also be argued that as the fix is available in a later release, it should provide incentive to move to that release.

It can also be beneficial if the post release BZ updater, checked the version and then closed the bug (which can be done: https://github.com/gluster/release-tools/blob/master/close-bugs.sh ).

If the above sounds fine, we can go with the change as is, OTOH it looks like one more rule to remember :) currently it is clone always, now it is clone when backporting (not too bad a rule to remember, but just stating it here.

Eitherway seems fine, not having to clone is lesser wasteful work, does not seem to loose too much in tracking information (if we address not closing the bug).

Reply:
I guess I didn't explain it right. Here is the flow I am envisioning. (and that is what this patch should achieve).

I see two type of bugs: 1) critical bugs which need backport, 2) non-critical bugs which doesn't need backport.

Chronological order of events for non-critical bug:

* Day0:  User files bug `bz312` which exists in release 3.12 branch (he is using it now).
* Day13:  Developer fixes 'bz312' in master (no need to clone here). submit the patch for the bug.
  * User gets notified about the fix being submitted / merged in master.
* Day 16-45: release branch happens
* Day 45-60: release is announced and the script to automatically close the `bz312` closes the bug with Fixed-In-Version as the current release version, which can be even `4.1.0`

Now the flow for critical fix:

* Day 0:  User files bug `bz312` which exists in release 3.12 branch (he is using it now).
* Day 5-30 :  Developer fixes 'bz312' in master (no need to clone here). submit the patch for the bug.
  * Users gets notified about the fix being submitted / merged in master.
* Day 6-31 (day of merging or next day): Developer clones the bug as it is critical (or in ideal world auto created when we say `backport to 3.12` in gerrit). This bug can be `bz378`, and developer submits the patch for this new bug in the release branch.
* Day 7 - 30: maintenance release of the release branch happens. The script to close the bugs automatically closes the bug `bz378`
* Day 16-45: release branch happens on master.
* Day 45-60: release is announced and the script to automatically close the bug closes the `bz312` with Fixed-In-Version as the current release version.

Hope this is clear. In this way, we don't have to bother about the 'Version' of bug on master branch.

Reply:
Ack! Let's use this, it works overall.